### PR TITLE
fix(cli): add `EMFILE` error handling for MacOS to clarify why Metro crashed

### PR DESCRIPTION
**Pull request by byCedric**
# Why

Fixes ENG-13486

This adds basic error handling for `EMFILE` errors on MacOS, and outputs two separate messages.

1. When Watchman is not installed
watchman-not-installed

2. When Watchman is in a bad state
watchman-bad-state

## What do you mean "watchman is in a bad state"?

Well, it turns out that the [command that Metro is using to determine if watchman is installed](https://github.com/facebook/metro/blob/f2a130c324c50b5c586b2d7e36bf6be7a8e49795/packages/metro-file-map/src/lib/checkWatchmanCapabilities.js#L24-L29) and can be used, might cause a [segfault](https://github.com/facebook/watchman/issues/400).

image

> You can test this by running: `watchman shutdown-server && watchman list-capabilities --output-encoding=json --no-pretty --no-spawn`

It's basically related to the `--no-spawn`, when running the watchman server, everything works fine.

image

image

So, whenever we detect the EMFILE on MacOS, and we detect Watchman is installed -- we try to restart this server and [reset all file watchers](https://github.com/facebook/metro/blob/f2a130c324c50b5c586b2d7e36bf6be7a8e49795/docs/Troubleshooting.md?plain=1#L8). On the next start, this should get watchman in a working state and Metro to use watchman instead.

# How

- Added error handling through `process.on('uncaughtException', ..)`
- Added watchman restart / detection commands
- Added log to install watchman
- Added log to troubleshoot watchman, following Metro's own [watchman troubleshooting guide](https://github.com/facebook/metro/blob/f2a130c324c50b5c586b2d7e36bf6be7a8e49795/docs/Troubleshooting.md?plain=1#L8)

# Test Plan

On my mac, this reproduces it reliably:

- `brew install watchman`
- `npx create-expo@latest ./test-expo-emfile --template default@next`
- `cd ./test-expo-emfile`
- `watchman shutdown-server` (to trigger the bad state)
- `npx expo start`
  - Should see the bad state warning

To test the install watchman warning:

- `brew remove watchman`
- `npx expo start`
  - Should see the install watchman warning

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]
👏

File: packages/@expo/cli/src/utils/errors.ts [hunk 1]
```diff
@@ -79,3 +80,39 @@ export class UnimplementedError extends Error {
...
+      execSync('watchman shutdown-server', { stdio: 'ignore' });
+      // NOTE(cedric): this both starts the watchman server and resets all watchers
+      execSync('watchman watch-del-all', { stdio: 'ignore' });
+
+      warn(
+        'Watchman is installed but was likely not enabled when starting Metro, try starting your project again.\nIf this problem persists, follow the troubleshooting guide of Watchman: https://facebook.github.io/watchman/docs/troubleshooting'
+      );
+    } catch {
+      warn(
+        `Your MacOS system limit does not allow enough watchers for Metro, install Watchman instead. Learn more: https://facebook.github.io/watchman/docs/install`
```
Comment (Simek) [inline packages/@expo/cli/src/utils/errors.ts:107]:
```suggestion
        `Your macOS system limit does not allow enough watchers for Metro, install Watchman instead. Learn more: https://facebook.github.io/watchman/docs/install`
```

Naming nit.

File: packages/@expo/cli/src/utils/errors.ts [hunk 2]
```diff
@@ -79,3 +80,39 @@ export class UnimplementedError extends Error {
     this.name = 'UnimplementedError';
   }
 }
+
+/**
+ * Add additional information when EMFILE errors are encountered.
+ * These errors originate from Metro's FSEventsWatcher due to `fsevents` going over MacOS system limit.
+ * Unfortunately, these limits in MacOS are relatively low compared to an average React Native project.
```
Comment (Simek) [inline packages/@expo/cli/src/utils/errors.ts:87]:
```suggestion
 * Unfortunately, these limits in macOS are relatively low compared to an average React Native project.
```

File: packages/@expo/cli/src/utils/errors.ts [hunk 3]
```diff
@@ -79,3 +80,39 @@ export class UnimplementedError extends Error {
...
+ *
+ * @see https://github.com/expo/expo/issues/29083
+ * @see https://github.com/facebook/metro/issues/834
+ * @see https://github.com/fsevents/fsevents/issues/42#issuecomment-62632234
+ */
+function handleTooManyOpenFileErrors(error: any) {
+  // Only enable special logging when running on MacOS and are running into the `EMFILE` error
```
Comment (Simek) [inline packages/@expo/cli/src/utils/errors.ts:94]:
```suggestion
  // Only enable special logging when running on macOS and are running into the `EMFILE` error
```

**Pull request merged**